File: src/onnx9000/extensions/custom/vision_ops.ts
/** Implementation details and semantic operations. */
export function rotatedNms(
    boxes: number[][],
    scores: number[],
    iouThreshold: number,
    scoreThreshold: number = 0.0
): number[] {
    const boundedBoxes: number[][] = [];
    /** Implementation details and semantic operations. */
    for(const box of boxes) {
        const [cx, cy, w, h, angle] = box;
        const cosA = Math.abs(Math.cos(angle));
        const sinA = Math.abs(Math.sin(angle));

        const bw = w * cosA + h * sinA;
        const bh = w * sinA + h * cosA;

        boundedBoxes.push([
            cx - bw / 2, cy - bh / 2,
            cx + bw / 2, cy + bh / 2
        ]);
    }

    let indices: number[] = [];
    /** Implementation details and semantic operations. */
    for(let i = 0; i < scores.length; i++) {
        /** Implementation details and semantic operations. */
        if(scores[i] >= scoreThreshold) indices.push(i);
    }
    indices.sort((a, b) => scores[b] - scores[a]);

    const keep: number[] = [];
    /** Implementation details and semantic operations. */
    while(indices.length > 0) {
        const curr = indices.shift()!;
        keep.push(curr);

        const currBox = boundedBoxes[curr];
        indices = indices.filter(i => {
            const box = boundedBoxes[i];
            const x1 = Math.max(currBox[0], box[0]);
            const y1 = Math.max(currBox[1], box[1]);
            const x2 = Math.min(currBox[2], box[2]);
            const y2 = Math.min(currBox[3], box[3]);

            const interArea = Math.max(0, x2 - x1) * Math.max(0, y2 - y1);
            /** Implementation details and semantic operations. */
            if(interArea === 0) return true;

            const area1 = (currBox[2] - currBox[0]) * (currBox[3] - currBox[1]);
            const area2 = (box[2] - box[0]) * (box[3] - box[1]);
            const iou = interArea / (area1 + area2 - interArea);
            return iou <= iouThreshold;
        });
    }
    return keep;
}

/** Implementation details and semantic operations. */

export function gridSample(
    inputTensor: number[][][][],
    grid: number[][][][],
    mode: string = "bilinear",
    paddingMode: string = "zeros",
    alignCorners: boolean = false
): number[][][][] {
    const N = inputTensor.length;
    const C = N > 0 ? inputTensor[0].length : 0;
    const H = C > 0 ? inputTensor[0][0].length : 0;
    const W = H > 0 ? inputTensor[0][0][0].length : 0;

    const HOut = grid.length > 0 ? grid[0].length : 0;
    const WOut = HOut > 0 ? grid[0][0].length : 0;

    const out: number[][][][] = [];
    /** Implementation details and semantic operations. */
    for(let n = 0; n < N; n++) {
        const batch: number[][][] = [];
        /** Implementation details and semantic operations. */
        for(let c = 0; c < C; c++) {
            const channel: number[][] = [];
            /** Implementation details and semantic operations. */
            for(let h = 0; h < HOut; h++) {
                channel.push([...Array(WOut)].map(() => 0.0));
            }
            batch.push(channel);
        }
        out.push(batch);
    }

    /** Implementation details and semantic operations. */

    for(let n = 0; n < N; n++) {
        /** Implementation details and semantic operations. */
        for(let hOut = 0; hOut < HOut; hOut++) {
            /** Implementation details and semantic operations. */
            for(let wOut = 0; wOut < WOut; wOut++) {
                const [x, y] = grid[n][hOut][wOut];

                let ix: number, iy: number;
                /** Implementation details and semantic operations. */
                if(alignCorners) {
                    ix = ((x + 1) / 2) * (W - 1);
                    iy = ((y + 1) / 2) * (H - 1);
                } else {
                    ix = ((x + 1) * W - 1) / 2;
                    iy = ((y + 1) * H - 1) / 2;
                }

                /** Implementation details and semantic operations. */

                if(paddingMode === "border") {
                    ix = Math.max(0.0, Math.min(W - 1.0, ix));
                    iy = Math.max(0.0, Math.min(H - 1.0, iy));
                } else if (paddingMode === "reflection") {
                    /** Implementation details and semantic operations. */
                    if(ix < 0) ix = -ix;
                    /** Implementation details and semantic operations. */
                    if(ix >= W) ix = W - 1 - (ix - W);
                    /** Implementation details and semantic operations. */
                    if(iy < 0) iy = -iy;
                    /** Implementation details and semantic operations. */
                    if(iy >= H) iy = H - 1 - (iy - H);
                }

                /** Implementation details and semantic operations. */

                if(mode === "nearest") {
                    const ixN = Math.round(ix);
                    const iyN = Math.round(iy);
                    /** Implementation details and semantic operations. */
                    for(let c = 0; c < C; c++) {
                        /** Implementation details and semantic operations. */
                        if(ixN >= 0 && ixN < W && iyN >= 0 && iyN < H) {
                            out[n][c][hOut][wOut] = inputTensor[n][c][iyN][ixN];
                        }
                    }
                } else {
                    const ixNw = Math.floor(ix);
                    const iyNw = Math.floor(iy);
                    const ixNe = ixNw + 1;
                    const iySw = iyNw + 1;

                    const dx = ix - ixNw;
                    const dy = iy - iyNw;

                    /** Implementation details and semantic operations. */

                    for(let c = 0; c < C; c++) {
                        const vNw = (ixNw >= 0 && ixNw < W && iyNw >= 0 && iyNw < H) ? inputTensor[n][c][iyNw][ixNw] : 0.0;
                        const vNe = (ixNe >= 0 && ixNe < W && iyNw >= 0 && iyNw < H) ? inputTensor[n][c][iyNw][ixNe] : 0.0;
                        const vSw = (ixNw >= 0 && ixNw < W && iySw >= 0 && iySw < H) ? inputTensor[n][c][iySw][ixNw] : 0.0;
                        const vSe = (ixNe >= 0 && ixNe < W && iySw >= 0 && iySw < H) ? inputTensor[n][c][iySw][ixNe] : 0.0;

                        out[n][c][hOut][wOut] = (
                            vNw * (1 - dx) * (1 - dy) +
                            vNe * dx * (1 - dy) +
                            vSw * (1 - dx) * dy +
                            vSe * dx * dy
                        );
                    }
                }
            }
        }
    }
    return out;
}

/** Implementation details and semantic operations. */

export function roiAlign(
    inputTensor: number[][][][],
    rois: number[][],
    batchIndices: number[],
    outputHeight: number,
    outputWidth: number,
    spatialScale: number,
    samplingRatio: number,
    aligned: boolean = true
): number[][][][] {
    const numRois = rois.length;
    const C = inputTensor[0].length;
    const H = inputTensor[0][0].length;
    const W = inputTensor[0][0][0].length;

    const out: number[][][][] = [];
    /** Implementation details and semantic operations. */
    for(let rIdx = 0; rIdx < numRois; rIdx++) {
        const roiChannels: number[][][] = [];
        /** Implementation details and semantic operations. */
        for(let c = 0; c < C; c++) {
            const plane: number[][] = [];
            /** Implementation details and semantic operations. */
            for(let ph = 0; ph < outputHeight; ph++) {
                plane.push([...Array(outputWidth)].map(() => 0.0));
            }
            roiChannels.push(plane);
        }
        out.push(roiChannels);
    }

    const offset = aligned ? 0.5 : 0.0;

    /** Implementation details and semantic operations. */

    for(let rIdx = 0; rIdx < numRois; rIdx++) {
        const bIdx = batchIndices[rIdx];
        let [x1, y1, x2, y2] = rois[rIdx];

        x1 = x1 * spatialScale - offset;
        y1 = y1 * spatialScale - offset;
        x2 = x2 * spatialScale - offset;
        y2 = y2 * spatialScale - offset;

        const roiW = Math.max(x2 - x1, 1.0);
        const roiH = Math.max(y2 - y1, 1.0);

        const binSizeH = roiH / outputHeight;
        const binSizeW = roiW / outputWidth;

        const gridH = samplingRatio > 0 ? samplingRatio : Math.ceil(roiH / outputHeight);
        const gridW = samplingRatio > 0 ? samplingRatio : Math.ceil(roiW / outputWidth);

        /** Implementation details and semantic operations. */

        for(let c = 0; c < C; c++) {
            /** Implementation details and semantic operations. */
            for(let ph = 0; ph < outputHeight; ph++) {
                /** Implementation details and semantic operations. */
                for(let pw = 0; pw < outputWidth; pw++) {
                    let val = 0.0;
                    /** Implementation details and semantic operations. */
                    for(let iy = 0; iy < gridH; iy++) {
                        /** Implementation details and semantic operations. */
                        for(let ix = 0; ix < gridW; ix++) {
                            let y = y1 + ph * binSizeH + (iy + 0.5) * binSizeH / gridH;
                            let x = x1 + pw * binSizeW + (ix + 0.5) * binSizeW / gridW;

                            /** Implementation details and semantic operations. */

                            if(y < -1.0 || y > H || x < -1.0 || x > W) continue;

                            y = Math.max(0.0, Math.min(H - 1.0, y));
                            x = Math.max(0.0, Math.min(W - 1.0, x));

                            const yLow = Math.floor(y);
                            const xLow = Math.floor(x);
                            const yHigh = Math.min(yLow + 1, H - 1);
                            const xHigh = Math.min(xLow + 1, W - 1);

                            const dy = y - yLow;
                            const dx = x - xLow;

                            const w1 = (1 - dx) * (1 - dy);
                            const w2 = dx * (1 - dy);
                            const w3 = (1 - dx) * dy;
                            const w4 = dx * dy;

                            val += (
                                w1 * inputTensor[bIdx][c][yLow][xLow] +
                                w2 * inputTensor[bIdx][c][yLow][xHigh] +
                                w3 * inputTensor[bIdx][c][yHigh][xLow] +
                                w4 * inputTensor[bIdx][c][yHigh][xHigh]
                            );
                        }
                    }
                    out[rIdx][c][ph][pw] = val / (gridH * gridW);
                }
            }
        }
    }
    return out;
}

/** Implementation details and semantic operations. */

export function deformConv2d(
    x: number[][][][],
    weight: number[][][][],
    offset: number[][][][],
    mask: number[][][][] | null = null,
    bias: number[] | null = null,
    stride: number = 1,
    padding: number = 0,
    dilation: number = 1
): number[][][][] {
    const N = x.length;
    const inC = x[0].length;
    const inH = x[0][0].length;
    const inW = x[0][0][0].length;

    const outC = weight.length;
    const kH = weight[0][0].length;
    const kW = weight[0][0][0].length;

    const outH = Math.floor((inH + 2 * padding - dilation * (kH - 1) - 1) / stride) + 1;
    const outW = Math.floor((inW + 2 * padding - dilation * (kW - 1) - 1) / stride) + 1;

    const out: number[][][][] = [];
    /** Implementation details and semantic operations. */
    for(let n = 0; n < N; n++) {
        const batch: number[][][] = [];
        /** Implementation details and semantic operations. */
        for(let oc = 0; oc < outC; oc++) {
            const channel: number[][] = [];
            /** Implementation details and semantic operations. */
            for(let oh = 0; oh < outH; oh++) {
                channel.push([...Array(outW)].map(() => 0.0));
            }
            batch.push(channel);
        }
        out.push(batch);
    }

    /** Implementation details and semantic operations. */

    for(let n = 0; n < N; n++) {
        /** Implementation details and semantic operations. */
        for(let oc = 0; oc < outC; oc++) {
            /** Implementation details and semantic operations. */
            for(let oh = 0; oh < outH; oh++) {
                /** Implementation details and semantic operations. */
                for(let ow = 0; outW > ow; ow++) {
                    let val = bias ? bias[oc] : 0.0;
                    /** Implementation details and semantic operations. */
                    for(let ic = 0; ic < inC; ic++) {
                        /** Implementation details and semantic operations. */
                        for(let kh = 0; kh < kH; kh++) {
                            /** Implementation details and semantic operations. */
                            for(let kw = 0; kw < kW; kw++) {
                                const ih = oh * stride - padding + kh * dilation;
                                const iw = ow * stride - padding + kw * dilation;

                                const offsetIdxY = 2 * (kh * kW + kw);
                                const offsetIdxX = offsetIdxY + 1;

                                const dy = offset[n][offsetIdxY][oh][ow];
                                const dx = offset[n][offsetIdxX][oh][ow];

                                const sampledY = ih + dy;
                                const sampledX = iw + dx;

                                let mVal = 1.0;
                                /** Implementation details and semantic operations. */
                                if(mask) {
                                    const maskIdx = kh * kW + kw;
                                    mVal = mask[n][maskIdx][oh][ow];
                                }

                                /** Implementation details and semantic operations. */

                                if(sampledY >= 0 && sampledY < inH && sampledX >= 0 && sampledX < inW) {
                                    const y0 = Math.floor(sampledY);
                                    const x0 = Math.floor(sampledX);
                                    const y1 = Math.min(y0 + 1, inH - 1);
                                    const x1 = Math.min(x0 + 1, inW - 1);

                                    const wy = sampledY - y0;
                                    const wx = sampledX - x0;

                                    const v00 = x[n][ic][y0][x0];
                                    const v01 = x[n][ic][y0][x1];
                                    const v10 = x[n][ic][y1][x0];
                                    const v11 = x[n][ic][y1][x1];

                                    const sVal = (
                                        v00 * (1 - wy) * (1 - wx) +
                                        v01 * (1 - wy) * wx +
                                        v10 * wy * (1 - wx) +
                                        v11 * wy * wx
                                    );

                                    val += sVal * weight[oc][ic][kh][kw] * mVal;
                                }
                            }
                        }
                    }
                    out[n][oc][oh][ow] = val;
                }
            }
        }
    }
    return out;
}

File: src/onnx9000/extensions/custom/vision_ops.py
"""Module providing core logic and structural definitions."""

from typing import List, Any
import math


def rotated_nms(
    boxes: List[List[float]],
    scores: List[float],
    iou_threshold: float,
    score_threshold: float = 0.0,
) -> List[int]:
    """
    Very simplified rotated NMS. Real rotated NMS requires polygon intersection.
    For this native implementation, we'll bound the rotated boxes and apply standard NMS.
    """
    from onnx9000.extensions.custom.ops import nms

    bounded_boxes = []
    for cx, cy, w, h, angle in boxes:
        # compute bounding box
        cos_a = abs(math.cos(angle))
        sin_a = abs(math.sin(angle))

        bw = w * cos_a + h * sin_a
        bh = w * sin_a + h * cos_a

        bounded_boxes.append([cx - bw / 2, cy - bh / 2, cx + bw / 2, cy + bh / 2])

    return nms(bounded_boxes, scores, iou_threshold, score_threshold)


def grid_sample(
    input_tensor: List[List[List[List[float]]]],
    grid: List[List[List[List[float]]]],
    mode: str = "bilinear",
    padding_mode: str = "zeros",
    align_corners: bool = False,
) -> List[List[List[List[float]]]]:
    """
    N, C, H, W input and N, H_out, W_out, 2 grid
    """
    N = len(input_tensor)
    C = len(input_tensor[0]) if N > 0 else 0
    H = len(input_tensor[0][0]) if C > 0 else 0
    W = len(input_tensor[0][0][0]) if H > 0 else 0

    H_out = len(grid[0]) if len(grid) > 0 else 0
    W_out = len(grid[0][0]) if H_out > 0 else 0

    out = [
        [[[0.0 for _ in range(W_out)] for _ in range(H_out)] for _ in range(C)]
        for _ in range(N)
    ]

    for n in range(N):
        for h_out in range(H_out):
            for w_out in range(W_out):
                x, y = grid[n][h_out][w_out]

                if align_corners:
                    ix = (x + 1) / 2 * (W - 1)
                    iy = (y + 1) / 2 * (H - 1)
                else:
                    ix = ((x + 1) * W - 1) / 2
                    iy = ((y + 1) * H - 1) / 2

                if padding_mode == "border":
                    ix = max(0.0, min(W - 1.0, ix))
                    iy = max(0.0, min(H - 1.0, iy))
                elif padding_mode == "reflection":
                    # Simplified reflection
                    if ix < 0:
                        ix = -ix
                    if ix >= W:
                        ix = W - 1 - (ix - W)
                    if iy < 0:
                        iy = -iy
                    if iy >= H:
                        iy = H - 1 - (iy - H)

                if mode == "nearest":
                    ix_n = int(round(ix))
                    iy_n = int(round(iy))
                    for c in range(C):
                        if 0 <= ix_n < W and 0 <= iy_n < H:
                            out[n][c][h_out][w_out] = input_tensor[n][c][iy_n][ix_n]
                else:  # bilinear
                    ix_nw = int(math.floor(ix))
                    iy_nw = int(math.floor(iy))
                    ix_ne = ix_nw + 1
                    iy_sw = iy_nw + 1

                    dx = ix - ix_nw
                    dy = iy - iy_nw

                    for c in range(C):
                        v_nw = (
                            input_tensor[n][c][iy_nw][ix_nw]
                            if (0 <= ix_nw < W and 0 <= iy_nw < H)
                            else 0.0
                        )
                        v_ne = (
                            input_tensor[n][c][iy_nw][ix_ne]
                            if (0 <= ix_ne < W and 0 <= iy_nw < H)
                            else 0.0
                        )
                        v_sw = (
                            input_tensor[n][c][iy_sw][ix_nw]
                            if (0 <= ix_nw < W and 0 <= iy_sw < H)
                            else 0.0
                        )
                        v_se = (
                            input_tensor[n][c][iy_sw][ix_ne]
                            if (0 <= ix_ne < W and 0 <= iy_sw < H)
                            else 0.0
                        )

                        out[n][c][h_out][w_out] = (
                            v_nw * (1 - dx) * (1 - dy)
                            + v_ne * dx * (1 - dy)
                            + v_sw * (1 - dx) * dy
                            + v_se * dx * dy
                        )
    return out


def roi_align(
    input_tensor: List[List[List[List[float]]]],
    rois: List[List[float]],
    batch_indices: List[int],
    output_height: int,
    output_width: int,
    spatial_scale: float,
    sampling_ratio: int,
    aligned: bool = True,
) -> List[List[List[List[float]]]]:
    """Provides semantic functionality and verification."""
    # Simplified RoIAlign
    num_rois = len(rois)
    C = len(input_tensor[0])
    H = len(input_tensor[0][0])
    W = len(input_tensor[0][0][0])

    out = [
        [
            [[0.0 for _ in range(output_width)] for _ in range(output_height)]
            for _ in range(C)
        ]
        for _ in range(num_rois)
    ]

    offset = 0.5 if aligned else 0.0

    for r_idx in range(num_rois):
        b_idx = batch_indices[r_idx]
        x1, y1, x2, y2 = rois[r_idx]

        x1 = x1 * spatial_scale - offset
        y1 = y1 * spatial_scale - offset
        x2 = x2 * spatial_scale - offset
        y2 = y2 * spatial_scale - offset

        roi_w = max(x2 - x1, 1.0)
        roi_h = max(y2 - y1, 1.0)

        bin_size_h = roi_h / output_height
        bin_size_w = roi_w / output_width

        grid_h = (
            sampling_ratio
            if sampling_ratio > 0
            else int(math.ceil(roi_h / output_height))
        )
        grid_w = (
            sampling_ratio
            if sampling_ratio > 0
            else int(math.ceil(roi_w / output_width))
        )

        for c in range(C):
            for ph in range(output_height):
                for pw in range(output_width):
                    val = 0.0
                    for iy in range(grid_h):
                        for ix in range(grid_w):
                            y = y1 + ph * bin_size_h + (iy + 0.5) * bin_size_h / grid_h
                            x = x1 + pw * bin_size_w + (ix + 0.5) * bin_size_w / grid_w

                            if y < -1.0 or y > H or x < -1.0 or x > W:
                                val += 0.0
                                continue

                            y = max(0.0, min(H - 1.0, y))
                            x = max(0.0, min(W - 1.0, x))

                            y_low = int(y)
                            x_low = int(x)
                            y_high = min(y_low + 1, H - 1)
                            x_high = min(x_low + 1, W - 1)

                            dy = y - y_low
                            dx = x - x_low

                            w1 = (1 - dx) * (1 - dy)
                            w2 = dx * (1 - dy)
                            w3 = (1 - dx) * dy
                            w4 = dx * dy

                            val += (
                                w1 * input_tensor[b_idx][c][y_low][x_low]
                                + w2 * input_tensor[b_idx][c][y_low][x_high]
                                + w3 * input_tensor[b_idx][c][y_high][x_low]
                                + w4 * input_tensor[b_idx][c][y_high][x_high]
                            )

                    out[r_idx][c][ph][pw] = val / (grid_h * grid_w)

    return out


def deform_conv2d(
    x: Any,
    weight: Any,
    offset: Any,
    mask: Any = None,
    bias: Any = None,
    stride: int = 1,
    padding: int = 0,
    dilation: int = 1,
) -> Any:
    """
    Naive DeformConv2D fallback.
    """
    N = len(x)
    in_C = len(x[0])
    in_H = len(x[0][0])
    in_W = len(x[0][0][0])

    out_C = len(weight)
    kH = len(weight[0][0])
    kW = len(weight[0][0][0])

    out_H = (in_H + 2 * padding - dilation * (kH - 1) - 1) // stride + 1
    out_W = (in_W + 2 * padding - dilation * (kW - 1) - 1) // stride + 1

    out = [
        [[[0.0 for _ in range(out_W)] for _ in range(out_H)] for _ in range(out_C)]
        for _ in range(N)
    ]

    for n in range(N):
        for oc in range(out_C):
            for oh in range(out_H):
                for ow in range(out_W):
                    val = bias[oc] if bias else 0.0
                    for ic in range(in_C):
                        for kh in range(kH):
                            for kw in range(kW):
                                # standard conv index
                                ih = oh * stride - padding + kh * dilation
                                iw = ow * stride - padding + kw * dilation

                                # add offset
                                # offset shape: N, 2*kH*kW, out_H, out_W
                                offset_idx_y = 2 * (kh * kW + kw)
                                offset_idx_x = offset_idx_y + 1

                                dy = offset[n][offset_idx_y][oh][ow]
                                dx = offset[n][offset_idx_x][oh][ow]

                                sampled_y = ih + dy
                                sampled_x = iw + dx

                                # apply mask if DCNv2
                                m_val = 1.0
                                if mask is not None:
                                    mask_idx = kh * kW + kw
                                    m_val = mask[n][mask_idx][oh][ow]

                                # bilinear sample
                                if 0 <= sampled_y < in_H and 0 <= sampled_x < in_W:
                                    y0 = int(math.floor(sampled_y))
                                    x0 = int(math.floor(sampled_x))
                                    y1 = min(y0 + 1, in_H - 1)
                                    x1 = min(x0 + 1, in_W - 1)

                                    wy = sampled_y - y0
                                    wx = sampled_x - x0

                                    v00 = x[n][ic][y0][x0]
                                    v01 = x[n][ic][y0][x1]
                                    v10 = x[n][ic][y1][x0]
                                    v11 = x[n][ic][y1][x1]

                                    s_val = (
                                        v00 * (1 - wy) * (1 - wx)
                                        + v01 * (1 - wy) * wx
                                        + v10 * wy * (1 - wx)
                                        + v11 * wy * wx
                                    )

                                    val += s_val * weight[oc][ic][kh][kw] * m_val
                    out[n][oc][oh][ow] = val

    return out
